### ISSUE TITLE
Connection Thread-Safety Example Request

### ISSUE THREAD

**Issue by gerlowskija** [mp911de/lettuce#480]
(Using lettuce 4.3.1.Final on Ubuntu 14.04)

This issue is opened more as a "question", please tag it as such.

The `lettuce` documentation clarifies that connection objects are thread-safe.  But a lot of the examples I cobble together seem to be unsafe without additional locking.  Particularly those involving multi/exec transactions.  As an example:

```
class WorkerThread extends Thread {
    private final StatefulRedisConnection<String, String> connection;

    public LettuceThread(StatefulRedisConnection<String, String> connection) {
        this.connection = connection;
    }

    @Override
    public void run() {
        final RedisCommands<String, String> sync = connection.sync();
        while (true) {
            sync.multi();
            sync.set("key", "value");
            sync.get("key");
            sync.exec();
        }
    }
}
```

The above example has concurrency issues, as different threads share the same synchronous RedisCommands object, and can accidentally interleave calls to `sync.multi()`, producing the error: `ERR MULTI calls can not be nested`.

Similar issues can be encountered if separate threads interleave transactional and non-transactional commands as well.

This is easy to fix with some locking and isn't a big deal, but it leaves me paranoid that I'm using the library incorrectly.  I'm not sure how to square the behavior with the thread-safety expressed in the documentation.

Is there a more conventional use of the library that avoids these concurrency pitfalls?  Is `lettuce` only thread-safe when multi/exec are avoided?  Are there any examples in the documentation that describe how to correctly use `lettuce` in a multi-threaded environment?

**Comment by mp911de** [comment]
Hi, thanks for your question. [StackOverflow](https://stackoverflow.com/questions/tagged/lettuce) is a great place to share questions for a greater audience.

Lettuce is thread-safe but transactions are not. Sharing connections that use transactions create a shared mutable state and that's something you want to avoid:

From the [Wiki](https://github.com/mp911de/lettuce/wiki):

> Lettuce is a scalable thread-safe Redis client providing synchronous, asynchronous and reactive APIs. Multiple threads may share one connection if they avoid blocking and transactional operations such as BLPOP and MULTI/EXEC

A good way how to deal with transactions is assigning connections directly to worker threads, or, use connection pooling if you use dynamic workers (such as handling web requests).

**Comment by gerlowskija** [comment]
I missed that caveat in the documentation.  Makes sense given what my team had observed.

Thanks for the clarification, and the nice library!

**Comment by phiSgr** [comment]
Sorry for commenting on an old thread.
I disagree with the description that "lettuce is thread-safe but transactions are not".

From that I expected:
> I should not use transactions with multiple threads, but lettuce is thread safe. So the worst that can happen is some exceptions thrown, but no internal client state corrupted.

I would rephrase it as "lettuce is thread-safe except for the transactions part of it". I phrase it this way because the `multi` variable is not protected by any locks (not that I want one).

Because of this misunderstanding I managed to make the Redis server accumulate millions of commands in a `MULTI` context. Finally I used a script to replace the transaction code.
